server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java

```diff
@@ -455,6 +560,40 @@ public Query nullValueQuery() {
             return termQuery(nullValue(), null);
         }
 
+        @Override
+        public Query analyzePhrase(String field, TokenStream stream, int slop) throws IOException {
+
+            if (indexPhrases && slop == 0 && hasGaps(stream) == false) {
+                stream = new FixedShingleFilter(stream, 2);
+                field = field + FAST_PHRASE_SUFFIX;
+            }
```

Comment:
I'd like this to work on all token streams and not only cached token streams, so maybe we should do something like:

```java
if (indexPhrases && slop == 0) {
  CachingTokenFilter cachedStream;
  if (stream instanceof CachingTokenFilter) {
    cachedStream = (CachingTokenFilter) stream;
  } else {
    stream = cachedStream = new CachedTokenStrem(stream);
  }
  if (hasGaps(cachedStream) == false) {
    stream = new FixedShingleFilter(stream, 2);
    field = field + FAST_PHRASE_SUFFIX;
  }
}
```
and change the signature of `hasGaps` to take a `CachingTokenFilter`?